ui/dashboards/src/utils/panelUtils.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { PanelDefinition } from '@perses-dev/core';
+import { PanelGroupDefinition } from '../context';
+
+// Given a PanelGroup, will find the Y coordinate for adding a new row to the grid, taking into account the items present
+export function getYForNewRow(group: PanelGroupDefinition) {
+  let newRowY = 0;
+  for (const layout of group.itemLayouts) {
+    const itemMaxY = layout.y + layout.h;
+    if (itemMaxY > newRowY) {
+      newRowY = itemMaxY;
+    }
+  }
+  return newRowY;
+}
+
+/**
+ * Get a valid panel key, where a valid key:
+ * - does not include invalid characters
+ * - is unique
+ */
+export function getValidPanelKey(panelKey: string, panels: Record<string, PanelDefinition>): string {
+  const uniquePanelKeys = getUniquePanelKeys(panels);
+  let normalizedPanelKey = getPanelKeyParts(removeWhiteSpacesAndSpecialCharacters(panelKey)).name;
+
+  const matchingKey = uniquePanelKeys[normalizedPanelKey];
+  if (typeof matchingKey === 'number') {
+    normalizedPanelKey += `-${matchingKey + 1}`;
+  }
+  return normalizedPanelKey;
+}
+
+type PanelKeyParts = {
+  name: string;
+  number?: number;
+};
+
+const removeWhiteSpacesAndSpecialCharacters = (str: string) => {
```

Comment:
This helper was kept from when `functions.ts` got renamed, but otherwise hasn't changed. 

Should we rename this to just `removeWhiteSpaces`? Maybe I'm missing some historical context?

Reply:
yep can rename it to `removeWhiteSpaces`